pghoard/basebackup/chunks.py

```diff
@@ -0,0 +1,328 @@
+# Copyright (c) 2022 Aiven, Helsinki, Finland. https://aiven.io/
+import hashlib
+import logging
+import os
+import socket
+import tarfile
+import threading
+import time
+from concurrent.futures import Future, ThreadPoolExecutor
+from dataclasses import dataclass, field
+from pathlib import Path
+from queue import Empty
+from tempfile import NamedTemporaryFile
+from types import TracebackType
+from typing import Any, Callable, Dict, List, Optional, Tuple, Type, Union
+
+from rohmu import rohmufile
+from rohmu.delta.common import EMBEDDED_FILE_SIZE
+
+# pylint: disable=superfluous-parens
+from pghoard.common import (
+    BackupFailure, BaseBackupFormat, CallbackEvent, CallbackQueue, CompressionData, EncryptionData, FileType, NoException
+)
+from pghoard.metrics import Metrics
+from pghoard.transfer import TransferQueue, UploadEvent
+
+
+class HashFile:
+    # Note when seek is used, hash will be wrong, we would need to implement re-read in case of seek is called which
+    # is a bit of overhead.
+    def __init__(self, *, path: Union[str, Path]) -> None:
+        self._file = open(path, "rb")
+        self.hash = hashlib.blake2s()
+
+    def __enter__(self) -> "HashFile":
+        return self
+
+    def __exit__(self, t: Optional[Type[BaseException]], v: Optional[BaseException], tb: Optional[TracebackType]) -> None:
+        self._file.close()
+
+    def read(self, n: Optional[int] = None) -> bytes:
+        data = self._file.read(n)
+        self.hash.update(data)
+        return data
+
+    def __getattr__(self, attr: str) -> Any:
+        return getattr(self._file, attr)
+
+
+@dataclass
+class DeltaStats:
+    hexdigests_sizes: Dict[str, int] = field(default_factory=dict)
+    # As some delta files technically might end up with the same hash, calculate total size separately
+    delta_files_total_size: int = 0
+    delta_files_total_count: int = 0
+    embed_count: int = 0
+    embed_size: int = 0
+    chunked_files_count: int = 0
+    chunked_files_size: int = 0
+
+
+class ChunkUploader:
+    def __init__(
+        self, metrics: Metrics, chunks_on_disk: int, encryption_data: EncryptionData, compression_data: CompressionData,
+        site_config: Dict[str, Any], site: str, is_running: Callable[[], bool], transfer_queue: TransferQueue
+    ):
+        self.log = logging.getLogger("ChunkUploader")
+        self.metrics = metrics
+        self.chunks_on_disk = chunks_on_disk
+        self.encryption_data = encryption_data
+        self.compression_data = compression_data
+        self.site_config = site_config
+        self.site = site
+        self.is_running = is_running
+        self.transfer_queue = transfer_queue
+        self.stats_lock = threading.Lock()
+
+    @staticmethod
+    def chunk_path_to_middle_path_name(chunk_path: Path, file_type: FileType) -> Tuple[Path, str]:
+        chunk_rel_path = chunk_path.relative_to(chunk_path.parent.parent)
+        if file_type == FileType.Basebackup_chunk:
```

Comment:
You should probably use pghoard.common.FileTypePrefixes here to avoid repeating those constants.